x-pack/qa/kerberos-tests/src/test/java/org/elasticsearch/xpack/security/authc/kerberos/CustomHttpClientConfigCallbackHandler.java

```diff
@@ -0,0 +1,338 @@
+     */
+    private static class KrbCallbackHandler implements CallbackHandler {
+        private final String principal;
+        private final SecureString password;
+
+        KrbCallbackHandler(final String principal, final SecureString password) {
+            this.principal = principal;
+            this.password = password;
+        }
+
+        public void handle(final Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+            for (Callback callback : callbacks) {
+                if (callback instanceof PasswordCallback) {
+                    PasswordCallback pc = (PasswordCallback) callback;
+                    if (pc.getPrompt().contains(principal)) {
+                        pc.setPassword(password.getChars());
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Usually we would have a JAAS configuration file for login configuration.
+     * Instead of an additional file setting as we do not want the options to be
+     * customizable we are constructing it in memory.
+     * <p>
+     * As we are using this instead of jaas.conf, this requires refresh of
+     * {@link Configuration} and reqires appropriate security permissions to do so.
+     */
+    private static class PasswordJaasConf extends AbstractJaasConf {
+
+        PasswordJaasConf(final String userPrincipalName, final Boolean enableDebugLogs) {
+            super(userPrincipalName, enableDebugLogs);
+        }
+
+        public void addOptions(final Map<String, String> options) {
+            options.put("useTicketCache", Boolean.FALSE.toString());
+            options.put("useKeyTab", Boolean.FALSE.toString());
+        }
+    }
+
+    /**
+     * Usually we would have a JAAS configuration file for login configuration. As
+     * we have static configuration except debug flag, we are constructing in
+     * memory. This avoids additional configuration required from the user.
+     * <p>
+     * As we are using this instead of jaas.conf, this requires refresh of
+     * {@link Configuration} and requires appropriate security permissions to do so.
+     */
+    private static class KeytabJaasConf extends AbstractJaasConf {
+        private final String keytabFilePath;
+
+        KeytabJaasConf(final String userPrincipalName, final String keytabFilePath, final Boolean enableDebugLogs) {
+            super(userPrincipalName, enableDebugLogs);
+            this.keytabFilePath = keytabFilePath;
+        }
+
+        public void addOptions(final Map<String, String> options) {
+            options.put("useKeyTab", Boolean.TRUE.toString());
+            options.put("keyTab", keytabFilePath);
+            options.put("doNotPrompt", Boolean.TRUE.toString());
+        }
+
+    }
+
+    private abstract static class AbstractJaasConf extends Configuration {
+        private final String userPrincipalName;
+        private final Boolean enableDebugLogs;
```

Comment:
why is this a `Boolean` and not a `boolean`? This could allow someone to pass in null which we do not want